bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+]
+"""
+
+from __future__ import annotations
+from dataclasses import dataclass
+from difflib import SequenceMatcher
+
+import json
+import os
+import requests
+import sys
+import tomllib
+
+from http.client import HTTPSConnection, HTTPResponse
+from typing import (
+    Any,
+    Dict,
+    List,
+    Generic,
+    Optional,
+    Tuple,
+    NamedTuple,
+    Set,
+    Iterable,
+    Protocol,
+    TypeVar,
+)
+class Member(NamedTuple):
+    id: str
+    name: str
+    email: str
+    type: str
+    access_all: Optional[bool]
+    # groups: Tuple[str, ...]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Member:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+        return Member(
+            id=data["member_id"],
+            name=data["member_name"],
+            email = data["email"],
+            type = data["type"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[member]]",
+            f"member_id = {self.id}",
+            f"member_name = {self.name}",
+            f"email = {self.email}",
+            f"type = {str(self.type)}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+class GroupMember(NamedTuple):
+    member_id: int
+    member_name: str
+    group_name: str
+
+    def get_id(self) -> str:
+        # Our generic differ has the ability to turn add/removes into changes
+        # for pairs with the same id, but this does not apply to memberships,
+        # which do not themselves have an id, so the identity to group on is
+        # the value itself.
+        return f"{self.member_id}@{self.group_name}"
+
+    def format_toml(self) -> str:
+        # Needed to satisfy Diffable, but not used in this case.
+        raise Exception(
+            "Group memberships are not expressed in toml, "
+            "please print the diffs in some other way."
+        )
+
+class Group(NamedTuple):
+    id: str
+    name: str
+    access_all: Optional[bool]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Group:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+
+        return Group(
+            id=data["group_id"],
+            name=data["group_name"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[group]]",
+            f"group_id = {self.id}",
+            f"group_name = {self.name}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+
+class MemberCollectionAccess(NamedTuple):
+    name: str
+
+    def get_id(self) -> str:
+        return self.id
```

Comment:
This field does not exist. Mypy should complain about this.

For context, the `get_id` function is used when diffing so we can distinguish between changing an object and deleting one and adding another. It’s always possible to represent a change as deleting the entire thing and then adding back an entirely new thing, but that would not generate a minimal diff. So we check if the thing removed and the thing added have the same id, and if they do, we show a change instead of add+remove.

Reply:
Yes you're right I missed those. Again copy and past bug sorry.. As you said `get_id` is only used for objects where we create a new `Diff`.

[6fd0242869](https://github.com/ChorusOne/github-access-manager/pull/10/commits/6fd0242869e02e934f6ec59fdb9c29a25f66e4ce)
```
❯ mypy --strict bitwarden_access_manager.py
bitwarden_access_manager.py:64: error: Cannot find implementation or library stub for module named "tomllib"  [import]
bitwarden_access_manager.py:64: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
bitwarden_access_manager.py:186: error: Returning Any from function declared to return "str"  [no-any-return]
bitwarden_access_manager.py:186: error: "GroupCollectionAccess" has no attribute "id"  [attr-defined]
bitwarden_access_manager.py:337: error: Function is missing a return type annotation  [no-untyped-def]
bitwarden_access_manager.py:339: error: "MemberCollectionAccess" has no attribute "id"  [attr-defined]
Found 5 errors in 1 file (checked 1 source file)
```